### ISSUE TITLE
streamCallbacks not fired

### ISSUE THREAD

**Issue by ohmer1** [flowplayer/flash#77]
I'm trying to catch UnpublishNotify event when the streamCallbacks function.  My function is never called.

I activated debug log and added a console.log in my callback function. My function is not called, but I see that flowplayer catch the wanted event ("_onNetStatus, code: NetStream.Play.UnpublishNotify")

Tried with flowplayer 3.2.15 with rtmp plugin 3.2.11 and flowplayer-3.2.16/rtmp-3.2.12.

Full debug output:

> [09:57:47,735] [INFO] time 09:57:47.734 :: org.flowplayer.rtmp::RTMPStreamProvider : _onNetStatus, code: NetStream.Play.UnpublishNotify
> [09:57:47,736] [LOG] time 09:57:47.736 :: org.flowplayer.rtmp::RTMPStreamProvider : using the default connection provider
> [09:57:47,737] [INFO] time 09:57:47.737 :: org.flowplayer.rtmp::RTMPStreamProvider : onNetStatus(), code: NetStream.Play.UnpublishNotify, paused? false, seeking? false
> [09:57:50,003] [INFO] time 09:57:50.02 :: org.flowplayer.rtmp::RTMPStreamProvider : _onNetStatus, code: NetStream.Buffer.Empty
> [09:57:50,004] [LOG] time 09:57:50.03 :: org.flowplayer.rtmp::RTMPStreamProvider : using the default connection provider
> [09:57:50,004] [LOG] time 09:57:50.04 :: org.flowplayer.rtmp::RTMPStreamProvider : silentSeek == false
> [09:57:50,005] [LOG] time 09:57:50.05 :: org.flowplayer.rtmp::RTMPStreamProvider : dispatching [ClipEvent type="onBufferEmpty" info=null] on clip [Clip] 'test1'
> [09:57:50,006] [INFO] time 09:57:50.06 :: org.flowplayer.rtmp::RTMPStreamProvider : onNetStatus(), code: NetStream.Buffer.Empty, paused? false, seeking? false`

Example:

``` javascript
    $f("live", "swf/flowplayer/flowplayer-3.2.15.swf", {

        log:
        {
          level: "debug",
          filter: "org.flowplayer.rtmp.*"
        },

        streamCallbacks: streamCallbacks: ["onUnpublishNotify"],

        clip:
        {  
            url: 'test1',
            live: true,
            provider: 'influxis',
            scaling: 'fit',
            accelerated: true,
            autoPlay: true,

            onUnpublishNotify: function(clip, info) {
              console.log("onUnpublishNotify() !!");
              alert("stream event triggered on clip " + clip.url + " info: " + info);
            }

        },    

        plugins: {
            influxis: {
                url: "swf/flowplayer/flowplayer.rtmp-3.2.11.swf",
                netConnectionUrl: 'rtmp://10.3.1.70/feed'
            }
        }
      });
```

**Comment by danrossi** [comment]
streamCallbacks: streamCallbacks: ["onUnpublishNotify"], are you sure this config is correct ? Have you tried

streamCallbacks:  ["onUnpublishNotify"],

**Comment by danrossi** [comment]
There seems to be a confusion with these events, these are custom callback events from the server, are you sure your server sends onUnpublishNotify ? It sounds like you want a standard unpublish event. I'll see if onNetStreamEvent will be suitable.

**Comment by danrossi** [comment]
looks like streamCallbacks handles custom methods, and onNetStreamEvent http://flash.flowplayer.org/documentation/events/clip.html#advanced-clip-events handles those specified callback methods. These need to be added in I think.

**Comment by danrossi** [comment]
Will possibly customise the rtmp plugin to dispatch all other events on this callback method I think this will get you going.

**Comment by danrossi** [comment]
Will send a onFinish event on unpublish events in the rtmp plugin.

**Comment by danrossi** [comment]
Hi there, there is a couple of issues with this, when you unpublish it's trying to reconnect to the stream and causes errors. I fixed this with HDS streams so it seems it needs this also. So once the stream is unpublished, it will now stop the player and you can use onStop: function() { }

https://dl.dropbox.com/u/3394987/flowplayer.rtmp-3.2.11.zip

**Comment by ohmer1** [comment]
Hello,

The fix works, I can call my function with the onStop function.

Problem is I don't know if the onStop is triggered by the unpublish event or if this is because the user clicked on the "pause" button.

**Comment by ohmer1** [comment]
Seem like the same problem occurs if the streaming server is shutted down.  The player receive a NetConnection.Connect.Closed event but the player is not stopped.  The image freeze, but onStop is not called.

**Comment by danrossi** [comment]
That is a secondary issue now. If you pause a stream onPause would be called right ? Im just following what jwplayer does and what ive added into the adobe http streaming plugin, it will call onStop when unpublish is called. What about onFinish  when disconnecting ?

**Comment by ohmer1** [comment]
When I click the "pause" button, only onStop() is fired.  onPause() is not.  I think onPause should be fired, but not onStop.

onFinish when disconnecting would be fine for me.

**Comment by danrossi** [comment]
there is no pause mode for a live stream it will trigger a stop event it has nothing to do with the fix. checking the disconnections now.

**Comment by danrossi** [comment]
I dont see why a production server would be shutting down in the middle of a broadcast, however im adding in the same method to triger onStop.

**Comment by danrossi** [comment]
try this https://dl.dropbox.com/u/3394987/flowplayer.rtmp-3.2.11.zip

**Comment by ohmer1** [comment]
Heh, I hope I won't have to shutdown the server in the middle of a broadcast.  But if it occurs, the player will react correctly. Thanks for the fix.

I understand there is no "pause" in live stream.  There is a way to detect if the stop is triggered by a user request or if the stop is because of a server event?

**Comment by danrossi** [comment]
There is no pause event for live streams it just stops so no.

**Comment by ohmer1** [comment]
Yes there is no pause.  I think the pause button should be replaced by a stop button for live stream.  Anyway, this is a detail.  Not really a problem.

Problem with the current implementation is that my onStop function display a message to a user saying he got disconnected from the server.  Then I try to reconnect it.  If the user click the pause/stop button to stop the stream, my onStop function fire and I reconnect the stream because I have no way to know if the stop is triggered by a user request of because of a server/network event.

Can you add a function parameter to onStop?  Could be a boolean, true if it's user requested and false for server/network event.

**Comment by danrossi** [comment]
If you want to deal with reconnections perhaps you might want to look at a custom plugin instead to manage this.

**Comment by danrossi** [comment]
leaving this open until release
